### PR TITLE
Fix AutoLinkPlugin triggers endless transforms

### LINKED ISSUE #5370: Bug: Converting to markdown breaks the editor with certain contents
Playing around with markdown conversion, I noticed that sometimes, for certain content, the conversion fails and breaks the editor entirely. 

Eg copying the contents of this [page](https://www.semrush.com/blog/free-seo-tools/) to the playground and trying to convert to markdown breaks the editor. Alternatively you can import this lexical state into the playground (I tried to attach a .lexical file but github doesn't let me): 
```
{"editorState":{"root":{"children":[{"children":[],"direction":null,"format":"","indent":0,"type":"paragraph","version":1},{"children":[{"detail":0,"format":0,"mode":"normal","style":"","text":"You can unlock much more keyword data in ","type":"text","version":1},{"children":[{"detail":0,"format":0,"mode":"normal","style":"","text":"Keyword Gap","type":"text","version":1}],"direction":"ltr","format":"","indent":0,"type":"link","version":1,"rel":null,"target":null,"title":null,"url":"https://www.semrush.com/analytics/keywordgap/"}],"direction":"ltr","format":"","indent":0,"type":"paragraph","version":1},{"children":[{"children":[{"detail":0,"format":0,"mode":"normal","style":"","text":"l","type":"text","version":1}],"direction":"ltr","format":"","indent":0,"type":"link","version":1,"rel":null,"target":null,"title":null,"url":"https://www.semrush.com/analytics/keywordgap/"}],"direction":"ltr","format":"","indent":0,"type":"paragraph","version":1},{"children":[{"detail":0,"format":0,"mode":"normal","style":"","text":"Keyword Manager","type":"text","version":1}],"direction":"ltr","format":"","indent":0,"type":"paragraph","version":1},{"children":[{"children":[{"detail":0,"format":0,"mode":"normal","style":"","text":"Keyword Manager","type":"text","version":1}],"direction":"ltr","format":"","indent":0,"type":"link","version":1,"rel":null,"target":null,"title":null,"url":"https://www.semrush.com/analytics/keywordmanager/"}],"direction":"ltr","format":"","indent":0,"type":"paragraph","version":1}],"direction":"ltr","format":"","indent":0,"type":"root","version":1}},"lastSaved":1702327856275,"source":"Playground","version":"0.12.5"}
```

I think there might be multiple situations in which something like this occurs, since sometimes deleting random characters seems to fix it.

Locally it returns the error: 
```
Uncaught Error: One or more transforms are endlessly triggering additional transforms. May have encountered infinite recursion caused by transforms that have their preconditions too lose and/or conflict with each other.
```

I tried figuring out which node or pattern was causing this, links, whitespace, whatever, but failed.

Lexical version: 0.12.5

Note: I'm not related to the blog page itself, it's just a sample in which this error occurs.

## Steps To Reproduce

1. Paste the contents of  [page](https://www.semrush.com/blog/free-seo-tools/) into the lexical playground (or import a .lexical file with the trimmed version of the contents above)
2. Try to convert it to markdown
3. The editor will break

### LINKED ISSUE #5403: Bug: AutoLinkPlugin triggers breakdown with certain input
Lexical version: Latest (0.12.5)

## Steps To Reproduce

1. go to the Playground
2. make editor empty
3. copy and paste this snippet on the editor:
```
123456789012345612345678901234567[link1](https://lexica1.dev/)[link2](https://lexica2.dev/)12345678901234567[link3](https://lexica3.dev/) 
```

Link to code example: can reproduce it on the Playground.

## The current behavior

https://github.com/facebook/lexical/assets/40269597/5202e79b-182c-4e79-be86-5687ddab39c2

### TL;DR
- **`text.substring` trims the left-sided invalid match text already, yet we still use `invalidMatchEnd` to adjust offset.**
- **This is somewhat duplicative.**
------------------
1. `AutoLinkPlugin` registers a listener in `useAutoLink`:
https://github.com/facebook/lexical/blob/520af1d29a0dd01194592429b487f45bbb8caa42/packages/lexical-react/src/LexicalAutoLinkPlugin.ts#L435
2. The listener calls `handleLinkCreation` and it calls `isContentAroundIsValid`. This function checks whether the characters at certain positions are separators by regex matching:
https://github.com/facebook/lexical/blob/520af1d29a0dd01194592429b487f45bbb8caa42/packages/lexical-react/src/LexicalAutoLinkPlugin.ts#L271
3. The current issue stems from passing an incorrect index here, specifically due to `invalidMatchEnd`.
5. The issue happens when this tells an invalid one as a valid one mistakenly.
    For example: when the match is `https://lexica2.dev/)`
    스크린샷 2023-12-19 오전 4 27 08

7.  It proceeds the logic for the valid around in `createAutoLinkNode`
    - `currentNode` gets split https://github.com/facebook/lexical/blob/520af1d29a0dd01194592429b487f45bbb8caa42/packages/lexical-react/src/LexicalAutoLinkPlugin.ts#L193
    - and merged again    https://github.com/facebook/lexical/blob/520af1d29a0dd01194592429b487f45bbb8caa42/packages/lexical/src/LexicalNormalization.ts#L81
8. This makes it repeat everything from the start endlessly.
9. The way I see it:
    - **`text.substring` trims the left-sided invalid match text already, yet we still use `invalidMatchEnd` to adjust offset.**
    - **This is somewhat duplicative.**
    
## The expected behavior

AutoLinkPlugin doesn't break the app.

## Proposed fix

~Let `text.substring` trim the invalid text and remove `invalidMatchEnd`.~
--> **this spawns some other bugs** 

Pass the constant initial text to `isContentAroundIsValid` instead of text which is being trimmed

### PR DESCRIPTION
**Pull request by 2wheeh**
fixes #5403 
fixes #5370 

before: in the Issues above

after:

https://github.com/facebook/lexical/assets/40269597/790f451f-a29a-4799-a7a8-d9114b193404

### CODE REVIEW

**Comment by 2wheeh** [comment]
https://github.com/facebook/lexical/issues/5403#issuecomment-1864880203

after:

https://github.com/facebook/lexical/assets/40269597/61927cc4-80da-4436-ab3d-b3edd3a05d48

**Review by zurfyx — APPROVED** [review]
Thanks for the fix! I'm suprised we don't need the full text anymore...
